# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: dom/quota/ActorsParent.cpp
# Commit: 7b3ae7465c92
# Full Hash: 7b3ae7465c9281d5674172961acd71a53507e7df
# Author: Jan Varga <jvarga@mozilla.com>
# Date: 2024-10-13 21:18:59
# Regressor Bug: 1866402
# File Overlap Count: 1
# Description:
#   Bug 1866402 - Add QuotaManager::PrincipalMetadataToPrincipalInfo method; r=dom-storage-reviewers,jari
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D195570
# ==============================================================================

diff -r 092d5b1c73ba -r 7b3ae7465c92 dom/quota/ActorsParent.cpp
--- a/dom/quota/ActorsParent.cpp	Sun Oct 13 08:24:48 2024 +0000
+++ b/dom/quota/ActorsParent.cpp	Sun Oct 13 10:10:23 2024 +0000
@@ -6566,6 +6566,27 @@
 }
 
 // static
+Result<PrincipalInfo, nsresult> QuotaManager::PrincipalMetadataToPrincipalInfo(
+    const PrincipalMetadata& aPrincipalMetadata) {
+  QM_TRY_INSPECT(
+      const auto& principal,
+      ([&aPrincipalMetadata]() -> Result<nsCOMPtr<nsIPrincipal>, nsresult> {
+        if (aPrincipalMetadata.mOrigin.EqualsLiteral(kChromeOrigin)) {
+          return nsCOMPtr<nsIPrincipal>(SystemPrincipal::Get());
+        }
+
+        return nsCOMPtr<nsIPrincipal>(
+            BasePrincipal::CreateContentPrincipal(aPrincipalMetadata.mOrigin));
+      }()));
+  QM_TRY(MOZ_TO_RESULT(principal));
+
+  PrincipalInfo principalInfo;
+  QM_TRY(MOZ_TO_RESULT(PrincipalToPrincipalInfo(principal, &principalInfo)));
+
+  return std::move(principalInfo);
+}
+
+// static
 nsAutoCString QuotaManager::GetOriginFromValidatedPrincipalInfo(
     const PrincipalInfo& aPrincipalInfo) {
   MOZ_ASSERT(IsPrincipalInfoValid(aPrincipalInfo));